### PR TITLE
crosscluster: add dlq tests and refactor

### PR DESCRIPTION
**Pull request by azhu-crl**
Followup PR to #126914. Changes included in this PR:

- Create dlq table with fully qualified prefixes instead of `defaultdb`
- Added unit tests to verify create dlq table & insert behave as expected
- Minor refactoring

Part of: https://cockroachlabs.atlassian.net/browse/DOC-10483
Epic: CRDB-38992
Release note: None

### CODE REVIEW

File: pkg/ccl/crosscluster/logical/dead_letter_queue.go
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/dead_letter_queue.go:137]: why is this necessary, given the create statement contains the `IF NOT EXISTS` clause? If the enum already exists, the statement should noop.

File: pkg/ccl/crosscluster/logical/dead_letter_queue.go
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/dead_letter_queue.go:142]: nit: i don't think it is necessary to include the tableID in the dlq table name

File: pkg/ccl/crosscluster/logical/dead_letter_queue.go
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/dead_letter_queue.go:153]: nit: `s/create-enum/create-dlq-enum/r`

File: pkg/ccl/crosscluster/logical/dead_letter_queue.go
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/dead_letter_queue.go:152]: to simplify things a bit, i think you should create the enum on the db.schema level, so in this sprintf, you can just pass the prefix.

File: pkg/ccl/crosscluster/logical/dead_letter_queue.go
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/dead_letter_queue.go:158]: ditto to the comment above: the enum should should live within the db's schema, so the last arg here could be prefix

File: pkg/ccl/crosscluster/logical/logical_replication_job.go
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/logical_replication_job.go:316]: nit: could you leave a todo here along the lines of: "combine tableIDToPrefixMap and dstToSrcDescMap"

File: pkg/ccl/crosscluster/replicationutils/utils.go
Comment (msbutler) [inline pkg/ccl/crosscluster/replicationutils/utils.go:252]: i think it's recomended to use `GetParentSchemaID()`. see note [here](https://github.com/msbutler/cockroach/blob/butler-skip-kms-roachtest/pkg/sql/catalog/descpb/structured.proto#L885)

File: pkg/ccl/crosscluster/logical/dead_letter_queue_test.go
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/dead_letter_queue_test.go:109]: could you create bar in a user defined schema? i.e. first create a schema and then create bar in it.

File: pkg/ccl/crosscluster/logical/dead_letter_queue_test.go
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/dead_letter_queue_test.go:171]: nite cleanup here!

**Review by msbutler — COMMENTED** [review]
Looks good! I think all my feedback should be quick to address.

File: pkg/ccl/crosscluster/replicationutils/utils.go
Comment (stevendanna) [inline pkg/ccl/crosscluster/replicationutils/utils.go:257]: _Somewhere_ I think we will want to escape each individual part of the name using `lexbase.EscapeSQLIdent`. Doing that here might make the parsing you are doing later harder.  See my comment on the protobuf for another idea.

File: pkg/ccl/crosscluster/logical/dead_letter_queue.go
Comment (stevendanna) [reply pkg/ccl/crosscluster/logical/dead_letter_queue.go:142]: Seems like below we are making a DLQ per table? But then we also are putting the table ID into the DLQ.

File: pkg/sql/execinfrapb/processors_bulk_io.proto
Comment (stevendanna) [inline pkg/sql/execinfrapb/processors_bulk_io.proto:518]: How about, rather than storing the pre-built prefix, we either store the entire descriptor or we store a new message type that has the database name and schema name as fields.  Then we wouldn't have to re-parse out the database name I think this would help with some other work we are doing as well.

File: pkg/ccl/crosscluster/logical/dead_letter_queue.go
Comment (stevendanna) [reply pkg/ccl/crosscluster/logical/dead_letter_queue.go:152]: I think @dt previously floated the idea of having a `crdb_replication` schema where we could put all of our types & tables. We could create that schema in every database.

**Review by stevendanna — COMMENTED** [review]
Thanks for driving this forward!

Left some comments, but defer to @msbutler for approval and review, don't wait on me.

File: pkg/ccl/crosscluster/logical/dead_letter_queue.go
Comment (msbutler) [reply pkg/ccl/crosscluster/logical/dead_letter_queue.go:152]: ah, i like that idea better

File: pkg/sql/execinfrapb/processors_bulk_io.proto
Comment (msbutler) [reply pkg/sql/execinfrapb/processors_bulk_io.proto:518]: works for me!

File: pkg/sql/execinfrapb/processors_bulk_io.proto
Comment (msbutler) [inline pkg/sql/execinfrapb/processors_bulk_io.proto:522]: why did you convert this into a slice? I think what steven and I were envisioning was:

```
type TableMetdata struct {
  sourcDescriptor descriptor
  databaseName string
}
```

then we keep this map here:
```
map[tableID]TableMetadata
```

File: pkg/sql/execinfrapb/processors_bulk_io.proto
Comment (msbutler) [inline pkg/sql/execinfrapb/processors_bulk_io.proto:485]: nit: i'd rename this to parent_database_name

also i think we still need to plumb the schema name field here, just to add to the dlq table _name_ to prevent name collisions. For example: suppose the user has the following tables: `mydb.schema_foo.baz` and `mydb.schema_bar.baz`, i think we want to create a unique dlq table for each with the table names: `schema_foo_baz` and `schema_bar_baz`.

File: pkg/ccl/crosscluster/logical/dead_letter_queue.go
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/dead_letter_queue.go:177]: nit: i don't think the tableID should be apart of the dlq table name, since the user doesn't need to know about it.

File: pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go:129]: nit: s/dstTableDescID/tableID/r

File: pkg/sql/execinfrapb/processors_bulk_io.proto
Comment (msbutler) [inline pkg/sql/execinfrapb/processors_bulk_io.proto:483]: also, we'll need to store the destination table name -- we can't rely on the source table name being the same as the destination name sadly.

File: pkg/ccl/crosscluster/logical/dead_letter_queue_test.go
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/dead_letter_queue_test.go:123]: nit: i'd add a test case with the schema: defaultDB.baz.foo

File: pkg/sql/execinfrapb/processors_bulk_io.proto
Comment (azhu-crl) [reply pkg/sql/execinfrapb/processors_bulk_io.proto:485]: re: dlq table name comments, should we settle on this naming convention: `db.crdb_replication.schema_table`?

Comment (msbutler) [reply pkg/sql/execinfrapb/processors_bulk_io.proto:485]: yeah, that's my preference

Comment (msbutler) [reply pkg/sql/execinfrapb/processors_bulk_io.proto:485]: yeah, that's my preference

Comment (msbutler) [reply pkg/sql/execinfrapb/processors_bulk_io.proto:485]: yeah, that's my preference

Comment (msbutler) [reply pkg/sql/execinfrapb/processors_bulk_io.proto:485]: yeah, that's my preference

File: pkg/ccl/crosscluster/logical/dead_letter_queue_test.go
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/dead_letter_queue_test.go:124]: could you add a test case that creates for a table not in `defaultdb`?

File: pkg/ccl/crosscluster/logical/dead_letter_queue_test.go
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/dead_letter_queue_test.go:156]: could you check that that the right enums were created in the right databases?

**Review by msbutler — APPROVED** [review]
Looks great! just left some small nits for the test that would be great to do before merging.

File: pkg/ccl/crosscluster/logical/logical_replication_job.go
Comment (stevendanna) [inline pkg/ccl/crosscluster/logical/logical_replication_job.go:353]: It isn't a _problem_ but this will end up trying to recreate the table on every single call of this function by the periodic plan checker.  A quick win here might be to add a flag to this function that we could pass in the replanner to avoid this extra work.

**Review by stevendanna — COMMENTED** [review]
Thanks for working on this! I've left some comments that I do think we'll want to address, but we can also address them as follow-up PRs if that is easier.

File: pkg/ccl/crosscluster/logical/dead_letter_queue.go
Comment (msbutler) [reply pkg/ccl/crosscluster/logical/dead_letter_queue.go:26]: So, I was envisioning that every schema change, like a rename, would trigger a replanning event, allowing us to create a new DLQ table for the renamed table. Old dlq events would remain in the old table. Are you saying that the user could connect the old and new dlq tables by noting the table id in the name? that seems reasonable.

File: pkg/ccl/crosscluster/logical/logical_replication_job.go
Comment (msbutler) [reply pkg/ccl/crosscluster/logical/logical_replication_job.go:353]: I think the logic would have to be a bit more complicated, given that I imagine schema changes (creating or renaming a table in the target) would require a new dlq table. Perhaps we can chat about this in -eng discussion today.

File: pkg/ccl/crosscluster/logical/dead_letter_queue.go
Comment (stevendanna) [reply pkg/ccl/crosscluster/logical/dead_letter_queue.go:26]: > So, I was envisioning that every schema change, like a rename, would trigger a replanning event

I think this is probably something close to what will happen, yeah.

> Are you saying that the user could connect the old and new dlq tables by noting the table id in the name? that seems reasonable.

Yeah.  But then, it does require the user to know the table ID, which is something we try not to do typically. I think let's leave it as is for now and we can look at this in more depth when we consider schema changes.

File: pkg/ccl/crosscluster/logical/logical_replication_job.go
Comment (stevendanna) [reply pkg/ccl/crosscluster/logical/logical_replication_job.go:353]: @msbutler Yes, but note that the plan made by the replan _checker_  never gets used. Rather it is just used to return an error. So my proposal is that the call made by the replan _checker_ would pass the "don't do extra work" flag and everything else would pass the "do all the work" flag.

Comment (msbutler) [reply pkg/ccl/crosscluster/logical/logical_replication_job.go:353]: ooooh, yeah you're right.

Comment (azhu-crl) [reply pkg/ccl/crosscluster/logical/logical_replication_job.go:353]: I'll need to look into the replanning logic a bit more, will address this and the comment above in another pr!

File: pkg/ccl/crosscluster/logical/dead_letter_queue_test.go
Comment (azhu-crl) [reply pkg/ccl/crosscluster/logical/dead_letter_queue_test.go:170]: tests are updated! :)

**Comment by azhu-crl** [comment]
TFTR!
bors r+

**Comment by azhu-crl** [comment]
blathers backport 24.2

**Pull request merged**
